internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/authzed/spicedb/internal/datastore"
+	"github.com/authzed/spicedb/internal/datastore/common"
+	"github.com/authzed/spicedb/internal/datastore/mysql/migrations"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/dlmiddlecote/sqlstats"
+	"github.com/go-sql-driver/mysql"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rs/zerolog/log"
+	"github.com/shopspring/decimal"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	colID               = "id"
+	colTimestamp        = "timestamp"
+	colNamespace        = "namespace"
+	colConfig           = "serialized_config"
+	colCreatedTxn       = "created_transaction"
+	colDeletedTxn       = "deleted_transaction"
+	colObjectID         = "object_id"
+	colRelation         = "relation"
+	colUsersetNamespace = "userset_namespace"
+	colUsersetObjectID  = "userset_object_id"
+	colUsersetRelation  = "userset_relation"
+
+	errRevision             = "unable to find revision: %w"
+	errCheckRevision        = "unable to check revision: %w"
+	errUnableToInstantiate  = "unable to instantiate datastore: %w"
+	errUnableToQueryTuples  = "unable to query tuples: %w"
+	errUnableToWriteTuples  = "unable to write tuples: %w"
+	errUnableToDeleteTuples = "unable to delete tuples: %w"
+	liveDeletedTxnID        = uint64(9223372036854775807)
+	batchDeleteSize         = 1000
+)
+
+var (
+	tracer = otel.Tracer("spicedb/internal/datastore/mysql")
+
+	// Unless specified otherwise, Go's MySQL driver will assume
+	// the server sends datetime in UTC,
+	// see https://github.com/go-sql-driver/mysql#loc. This parameter
+	// is unrelated to the session's timezone.
+	// If the server's global timezone is set to something other than UTC,
+	// the driver will incorrectly convert SELECT NOW(), because
+	// the default session timezone is the one specified by the server.
+	getNow = sb.Select("UTC_TIMESTAMP(6)")
+
+	sb = sq.StatementBuilder.PlaceholderFormat(sq.Question)
+)
+
+type sqlFilter interface {
+	ToSql() (string, []interface{}, error)
+}
+
+func NewMySQLDatastore(url string, options ...Option) (*Datastore, error) {
+	config, err := generateConfig(options)
+	if err != nil {
+		return nil, fmt.Errorf(errUnableToInstantiate, err)
+	}
+	connector, err := mysql.MySQLDriver{}.OpenConnector(url)
+	if err != nil {
+		return nil, fmt.Errorf("NewMySQLDatastore: failed to create connector: %w", err)
+	}
+	var db *sql.DB
+	if config.enablePrometheusStats {
+		connector, err = instrumentConnector(connector)
+		if err != nil {
+			return nil, fmt.Errorf("NewMySQLDatastore: unable to instrument connector: %w", err)
+		}
+		db = sql.OpenDB(connector)
+		// Create a new collector, the name will be used as a label on the metrics
```

Comment:
newline before this line

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/69f52043f21427f097ed76434b511c2bcc6b2ce3